### DISCUSSION TITLE
How to write a custom preset for milkdown

### DISCUSSION THREAD

**Discussion by navanshu** [Milkdown/milkdown#878] [Q&A]
Hello, I am trying to write [fountain](https://github.com/mattdaly/Fountain.js/blob/master/fountain.js) syntax with Milkdown, which means I have to replace preset-commonmark with custom parser, I was reading the code of preset-commonmark (v6 branch) and saw there are many components involved. 

It is divided into mark, node and plugin. Compared to commonmark, Fountain is straightforward syntax, so the implementation might be simpler. I have understood the working of a preset somewhat, but I will be glad to have a good understanding of it.

**Comment by Saul-Mirone** [comment]
I really recommend you to wait for the V7 releasing because there're breaking changes in plugins API. The release will happen at the middle of next month (Feb).

1. The factory API will be replaced with composable API. 
2. Milkdown will not provide style itself (which means it will be a headless component).

Before getting started, you can take a look into the code of preset-commonmark in main branch.

Basically:

1. You need to provide node and mark for different markdown syntax.
2. Use `inputrules` and `keymap` to provide abilities for users to create the node or mark.
3. Use `command` to provide abilities for developers to create your node or mark programmatically.

**Reply by navanshu** [reply]
Now I understand why the main branch has no theme in it. I really do like the UI of Milkdown and new website is also cool.

I am getting the feel of the api, as I have not ever built an editor, it's quite a new experience to me. I will explore the new api soon.

I wanted to recommend you to try a software "logseq". I like one feature it has is, when a block of text is clicked, it shows the exact markdown of that block. It improves the experience of working with links, bold, italics. Milkdown is just there, far ahead than other browser based markdown editors, its just lacking this feature.

**Reply by Saul-Mirone** [reply]
> Now I understand why the main branch has no theme in it. I really do like the UI of Milkdown and new website is also cool.
> 
> I am getting the feel of the api, as I have not ever built an editor, it's quite a new experience to me. I will explore the new api soon.
> 
> I wanted to recommend you to try a software "logseq". I like one feature it has is, when a block of text is clicked, it shows the exact markdown of that block. It improves the experience of working with links, bold, italics. Milkdown is just there, far ahead than other browser based markdown editors, its just lacking this feature.

As for theme, in fact the nord theme will still exists. There're no conflicts between headless component and theme. In fact I plan to add it back in a few commits.
Thanks for your recommendation. I'll take a look into logseq.

**Reply by navanshu** [reply]
Okay, I will wait for v7.
